keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -0,0 +1,223 @@
+- Defaulting bloats the resulting config, thus making it hard for users to migrate it upon a version change.
+
+Hence, kubeadm should stop defaulting component configs and leave this to components themselves.
+
+### Delegate config validation to the components
+
+Currently, kubeadm vendors in pieces of Go code from the components to be able to perform component config validation. This is a bad practice as it bloats the kubeadm dependencies and deviation between the vendored code and the components can lead to inconsistencies.  
+To solve this, validation is going to be performed by executing a container or the binary of the component to be used with an appropriate command to verify the supplied config.  
+If a target component does not have any means to verify its config, then kubeadm won’t do any validation for the time being. It will instead display a warning to the user denoting that certain component’s config is unvalidated.
+
+### Kubernetes Core vs AddOns Component Configs
+
+Currently kubeadm maintains the component configs of both kubelet and kube-proxy. These are components that are maintained in very different ways.
+
+Kubelet, along with API Server, Scheduler, Controller Manager and etcd, are deemed essential Kubernetes components. They are deployed and managed locally by kubeadm, using systemd services (kubelet) or static pods on the local kubelet (API Server, Scheduler, etc.). Hence, their configs are usually consumed by the components in the form of local files.
+
+On the other hand, kube-proxy and CoreDNS/kube-dns are addons, that are managed as standard Kubernetes deployments. Hence their configs are usually accessed via config maps.
+
+Therefore, two different config lifecycles are distinguishable here - one for Kubernetes core components and one for cluster addons.
+
+The defaulting and validation changes, proposed in the previous paragraphs, are going to be applied to both Kubernetes core and AddOn component config types.
+
+#### AddOn Component Config management
+
+AddOn component configs are always stored in config maps. They can be supplied in their entirety by users along with the kubeadm config during init or can be generated by kubeadm itself.
+If, upon upgrade, kubeadm detects an obsolete config version, it will bail out with an error message, asking users to upgrade the addon config by hand.
+
+The ultimate goal for addon component configs, is for them to be eventually managed in their entirety by cluster addon operators, instead of kubeadm.
+
+#### Kubernetes Core Component Config management
+
+Like in the addon case, Kubernetes core component configs can be specified by users in their entirety along with the kubeadm config during init or in a config map. On the other hand, if those configs are not user supplied, they are generated by kubeadm.
+User supplied configs need to be stored in config maps and have to be manually converted by users upon upgrade, while kubeadm generated ones need neither of these things.  
```

Comment:
shouldn't we rely on the components to do conversion for the users ?

Reply:
In general, that would be OK. The question is to what extent we can rely on the machine generated conversion? My gut feeling, is that machine generated conversion is the best.
However, on the component config brainstorming meeting (recording [here](https://www.youtube.com/watch?v=qThNvuhM-kY)) @liggitt and @mtaufen expressed an opinion, that we should regenerate configs, that are kubeadm originated and do nothing about user supplied ones.
So the original question of "Who is responsible for component config conversion? Is it cluster lifecycle tools or the components themselves?" was answered with "Don't do any automatic conversion. Expect the user to do that when necessary." and this KEP is in a direct response of that.

Reply:
Wondering is splitting config upgrade and component upgrade might create problems for the users.
e.g the user changes config before an upgrade, while the installed component is still running in an older version until upgrade is completed

Reply:
Depends on the component. Right now we have kubelet and kube-proxy and they won't be affected much.
For kubelet, you need to execute upgrade to apply the changes.
For kube-proxy, the CM is used at startup time for each kube-proxy instance. So all running instances ATM won't be affected. Only restarting instances will, but then I don't think, that loosing a kube-proxy instance for small amount of time during upgrade is that much of a concern.